R/FilterEnsemble.R

```diff
@@ -0,0 +1,296 @@
+        FUN = function(x) order(fval$data$value[x])))
+
+    fval_all_ranked_simple = fval_all_ranked_simple[with(fval_all_ranked_simple,
+      order(value, rank)), ]
+
+    ### calculate ensemble filter
+
+    # group by "name" and summarize the minimum of "rank"
+    fval_ens = aggregate(fval_all_ranked_simple$rank,
+      by = list(fval_all_ranked_simple$name), FUN = max)
+    colnames(fval_ens) = c("name", "value")
+
+    # add columns "type" and "method"
+    fval_ens$type = fval$data$type[1:length(unique(fval$data$name))]
+    fval_ens$method = "E-max"
+
+    # merge ensemble and base filters
+    fval_all_ranked_simple$rank = NULL
+    return(rbind(fval_all_ranked_simple, fval_ens))
+  }
+)
+
+# E-median ----------------
+#' Median ensemble filter. Takes the median across all base filter methods for
+#' each feature.
+#'
+#' @rdname makeFilter
+#' @name makeFilter
+makeFilterEnsemble(
+  name = "E-median",
+  desc = "Median ensemble filter. Takes the median across all base filter methods for each feature.",
+  base.methods = NULL,
+  fun = function(task, base.methods, nselect, more.args, ...) {
+
+    # calculate base filters here
+    fval = generateFilterValuesData(task, method = base.methods,
+      nselect = nselect, more.args = ...)
+
+    # rank base filters by method
+    fval_all_ranked_simple = transform(fval$data,
+      rank = ave(1:nrow(fval$data), method,
+        FUN = function(x) order(fval$data$value[x])))
+
+    fval_all_ranked_simple = fval_all_ranked_simple[with(fval_all_ranked_simple,
+      order(value, rank)), ]
+
+    ### calculate ensemble filter
+
+    # group by "name" and summarize the minimum of "rank"
+    fval_ens = aggregate(fval_all_ranked_simple$rank,
+      by = list(fval_all_ranked_simple$name), FUN = median)
+    colnames(fval_ens) = c("name", "value")
+
+    # add columns "type" and "method"
+    fval_ens$type = fval$data$type[1:length(unique(fval$data$name))]
+    fval_ens$method = "E-median"
+
+    # merge ensemble and base filters
+    fval_all_ranked_simple$rank = NULL
+    return(rbind(fval_all_ranked_simple, fval_ens))
+  }
+)
+
+# E-Borda ----------------
+#' Borda ensemble filter. Takes the sum across all base filter methods for each
+#' feature.
+#'
+#' @rdname makeFilter
+#' @name makeFilter
+makeFilterEnsemble(
+  name = "E-Borda",
+  desc = "Borda ensemble filter. Takes the sum across all base filter methods for each feature.",
+  base.methods = NULL,
+  fun = function(task, base.methods, nselect, more.args, ...) {
+
+    # calculate base filters here
+    fval = generateFilterValuesData(task, method = base.methods,
+      nselect = nselect, more.args = ...)
+
+    # rank base filters by method
+    fval_all_ranked_simple = transform(fval$data,
+      rank = ave(1:nrow(fval$data), method,
+        FUN = function(x) order(fval$data$value[x])))
+
+    fval_all_ranked_simple = fval_all_ranked_simple[with(fval_all_ranked_simple,
+      order(value, rank)), ]
```

Comment:
It looks like there's a lot of repeated code in these calls (for example the line above). Any way of moving this into the `makeFilterEnsemble` implementation?

Reply:
We cannot move it into `makeFilterEnsemble()` directly since S3 kicks in and causes troubles. But I moved the shared code into two helper funs. Good point.